# Tutorial: Joining Tracking and Play-by-Play Data
Pascale Walters, Mar. 26, 2022

## Step 1: Find the power play information

In [1]:
import pandas as pd

In [2]:
power_play_info_file = 'pp_info.csv'
power_play_info = pd.read_csv(power_play_info_file)
power_play_info.head()

game_name  penalty_number  start_video_clock_seconds  \
0  2022-02-08 Canada at USA               1                       1278   
1  2022-02-08 Canada at USA               2                       1624   
2  2022-02-08 Canada at USA               3                        409   
3  2022-02-08 Canada at USA               4                        898   
4  2022-02-08 Canada at USA               5                       1690   

   end_video_clock_seconds  start_period  end_period  \
0                     1314             1           1   
1                     1744             1           1   
2                      562             2           2   
3                      898             2           2   
4                     1927             2           2   

   start_game_clock_seconds  end_game_clock_seconds  
0                       386                     350  
1                       216                      96  
2                       991                     871  
3                       613                     613  
4                       272                     152

Let's focus on the first power play in the information file! It happened during the game between Canada and USA on 2022-02-08 and was the first power play of the game (`penalty_number = 1`). This penalty happened during the first period, and did not overflow into the second (`start_period = 1`, `end_period = 1`). The game clock read 386 seconds at the beginning, which is 6:26, and ended at 350 seconds, or 5:50.

## Step 2: Find the play-by-play data

Now that we know the start and end times of the given period, we can look into the play-by-play data to find the associated events.

In [4]:
play_by_play_data = pd.read_csv('../pxp_womens_oly_2022_v2.csv')
play_by_play_data.head()

game_date  season_year                        team_name  \
0  8/2/2022         2021         Olympic (Women) - Canada   
1  8/2/2022         2021         Olympic (Women) - Canada   
2  8/2/2022         2021         Olympic (Women) - Canada   
3  8/2/2022         2021         Olympic (Women) - Canada   
4  8/2/2022         2021  Olympic (Women) - United States   

                     opp_team_name venue  period  clock_seconds  \
0  Olympic (Women) - United States  away       1           1200   
1  Olympic (Women) - United States  away       1           1199   
2  Olympic (Women) - United States  away       1           1198   
3  Olympic (Women) - United States  away       1           1197   
4         Olympic (Women) - Canada  home       1           1194   

  situation_type  goals_for  goals_against  ... event_successful x_coord  \
0         5 on 5          0              0  ...                t     100   
1         5 on 5          0              0  ...                t      86   
2         5 on 5          0              0  ...                f      96   
3         5 on 5          0              0  ...                t     124   
4         5 on 5          0              0  ...                t       6   

  y_coord  event_type  player_name_2 x_coord_2 y_coord_2  event_detail_1  \
0      42    Backhand  Hannah Brandt       NaN       NaN             NaN   
1      31         NaN            NaN       NaN       NaN             NaN   
2      15         NaN            NaN       NaN       NaN             NaN   
3       1      Dumped   Cayla Barnes       NaN       NaN             NaN   
4      73         NaN            NaN       NaN       NaN             NaN   

   event_detail_2 event_detail_3  
0             NaN            NaN  
1             NaN            NaN  
2             NaN            NaN  
3             NaN            NaN  
4             NaN            NaN  

[5 rows x 22 columns]

Let's select the events for our game (2022-02-08 Canada at USA).

In [5]:
# Look for events that happened on 8/2/2022, and where the team performing the event is either Canada or USA
game_events = play_by_play_data.loc[(play_by_play_data['game_date'] == '8/2/2022') & 
                                    ((play_by_play_data['team_name'] == 'Olympic (Women) - Canada') |
                                        (play_by_play_data['team_name'] == 'Olympic (Women) - United States'))]
game_events.head()

game_date  season_year                        team_name  \
0  8/2/2022         2021         Olympic (Women) - Canada   
1  8/2/2022         2021         Olympic (Women) - Canada   
2  8/2/2022         2021         Olympic (Women) - Canada   
3  8/2/2022         2021         Olympic (Women) - Canada   
4  8/2/2022         2021  Olympic (Women) - United States   

                     opp_team_name venue  period  clock_seconds  \
0  Olympic (Women) - United States  away       1           1200   
1  Olympic (Women) - United States  away       1           1199   
2  Olympic (Women) - United States  away       1           1198   
3  Olympic (Women) - United States  away       1           1197   
4         Olympic (Women) - Canada  home       1           1194   

  situation_type  goals_for  goals_against  ... event_successful x_coord  \
0         5 on 5          0              0  ...                t     100   
1         5 on 5          0              0  ...                t      86   
2         5 on 5          0              0  ...                f      96   
3         5 on 5          0              0  ...                t     124   
4         5 on 5          0              0  ...                t       6   

  y_coord  event_type  player_name_2 x_coord_2 y_coord_2  event_detail_1  \
0      42    Backhand  Hannah Brandt       NaN       NaN             NaN   
1      31         NaN            NaN       NaN       NaN             NaN   
2      15         NaN            NaN       NaN       NaN             NaN   
3       1      Dumped   Cayla Barnes       NaN       NaN             NaN   
4      73         NaN            NaN       NaN       NaN             NaN   

   event_detail_2 event_detail_3  
0             NaN            NaN  
1             NaN            NaN  
2             NaN            NaN  
3             NaN            NaN  
4             NaN            NaN  

[5 rows x 22 columns]

Now that we have our game events, we can further filter them down to the period we want.

In [6]:
period_events = game_events.loc[game_events['period'] == 1]
period_events.head()

game_date  season_year                        team_name  \
0  8/2/2022         2021         Olympic (Women) - Canada   
1  8/2/2022         2021         Olympic (Women) - Canada   
2  8/2/2022         2021         Olympic (Women) - Canada   
3  8/2/2022         2021         Olympic (Women) - Canada   
4  8/2/2022         2021  Olympic (Women) - United States   

                     opp_team_name venue  period  clock_seconds  \
0  Olympic (Women) - United States  away       1           1200   
1  Olympic (Women) - United States  away       1           1199   
2  Olympic (Women) - United States  away       1           1198   
3  Olympic (Women) - United States  away       1           1197   
4         Olympic (Women) - Canada  home       1           1194   

  situation_type  goals_for  goals_against  ... event_successful x_coord  \
0         5 on 5          0              0  ...                t     100   
1         5 on 5          0              0  ...                t      86   
2         5 on 5          0              0  ...                f      96   
3         5 on 5          0              0  ...                t     124   
4         5 on 5          0              0  ...                t       6   

  y_coord  event_type  player_name_2 x_coord_2 y_coord_2  event_detail_1  \
0      42    Backhand  Hannah Brandt       NaN       NaN             NaN   
1      31         NaN            NaN       NaN       NaN             NaN   
2      15         NaN            NaN       NaN       NaN             NaN   
3       1      Dumped   Cayla Barnes       NaN       NaN             NaN   
4      73         NaN            NaN       NaN       NaN             NaN   

   event_detail_2 event_detail_3  
0             NaN            NaN  
1             NaN            NaN  
2             NaN            NaN  
3             NaN            NaN  
4             NaN            NaN  

[5 rows x 22 columns]

And now, we get the events that happened during our power play! Remember, we want events between 386 and 350 clock seconds.

In [7]:
pp_events = period_events.loc[(period_events['clock_seconds'] <= 386) &
                              (period_events['clock_seconds'] >= 350)]
pp_events.head()

game_date  season_year                        team_name  \
435  8/2/2022         2021  Olympic (Women) - United States   
436  8/2/2022         2021  Olympic (Women) - United States   
437  8/2/2022         2021  Olympic (Women) - United States   
438  8/2/2022         2021  Olympic (Women) - United States   
439  8/2/2022         2021         Olympic (Women) - Canada   

                       opp_team_name venue  period  clock_seconds  \
435         Olympic (Women) - Canada  home       1            386   
436         Olympic (Women) - Canada  home       1            386   
437         Olympic (Women) - Canada  home       1            383   
438         Olympic (Women) - Canada  home       1            382   
439  Olympic (Women) - United States  away       1            380   

    situation_type  goals_for  goals_against  ... event_successful x_coord  \
435         5 on 5          0              0  ...                f       5   
436         4 on 5          0              0  ...                t      31   
437         4 on 5          0              0  ...                t       6   
438         4 on 5          0              0  ...                f       4   
439         5 on 4          0              0  ...                t     165   

    y_coord      event_type        player_name_2 x_coord_2 y_coord_2  \
435      67  Cross-checking        Sarah Fillier       NaN       NaN   
436      65        Backhand  Marie-Philip Poulin       NaN       NaN   
437      64             NaN                  NaN       NaN       NaN   
438      37             NaN                  NaN       NaN       NaN   
439      82             NaN                  NaN       NaN       NaN   

     event_detail_1  event_detail_2 event_detail_3  
435             NaN             NaN            NaN  
436             NaN             NaN            NaN  
437             NaN             NaN            NaN  
438             NaN             NaN            NaN  
439             NaN             NaN            NaN  

[5 rows x 22 columns]

## Step 3: Get tracking data for the power play

This is a bit easier, since we have a dedicated file for the power play tracking.

In [8]:
pp_tracking = pd.read_csv('2022-02-08 Canada at USA/2022-02-08 Canada at USA P1 PP1.csv')
pp_tracking.head()

frame_id  period  track_id team_id team_name  jersey_number        x_ft  \
0       301       1         1   Light    Canada              6  171.071705   
1       301       1         2    Dark       USA              2  194.278571   
2       301       1         3    Dark       USA             37  194.258524   
3       301       1         4   Light    Canada             11  202.732672   
4       301       1         5    Dark       USA             28  174.116300   

        y_ft  
0  46.072340  
1  41.647639  
2  47.182641  
3  67.352362  
4  38.757970

## Step 4: Merge the play-by-play and tracking data

This is probably the most complicated part of the process! One **really** important thing to note: events are given at the second scale and tracking observations are given **30 times per second**.

How do we deal with this? It's kind of up to you and the question you're looking to answer! I'll give an example here.

Let's first take a look at the frame IDs given in the tracking data.

In [10]:
pp_tracking['frame_id'].unique()

array([ 301,  302,  303,  304,  305,  306,  307,  308,  309,  310,  311,
        312,  313,  314,  315,  316,  317,  318,  319,  320,  321,  322,
        323,  324,  325,  326,  327,  328,  329,  330,  331,  332,  333,
        334,  335,  336,  337,  338,  339,  340,  341,  342,  343,  344,
        345,  346,  347,  348,  349,  350,  351,  352,  353,  354,  355,
        356,  357,  358,  359,  360,  361,  362,  363,  364,  365,  366,
        367,  368,  369,  370,  371,  372,  373,  374,  375,  376,  377,
        378,  379,  380,  381,  382,  383,  384,  385,  386,  387,  388,
        389,  390,  391,  392,  393,  394,  395,  396,  397,  398,  399,
        400,  401,  402,  403,  404,  405,  406,  407,  408,  409,  410,
        411,  412,  413,  414,  415,  416,  417,  418,  419,  420,  421,
        422,  423,  424,  425,  426,  427,  428,  429,  430,  431,  432,
        433,  434,  435,  436,  437,  438,  439,  440,  441,  442,  443,
        444,  445,  446,  447,  448,  449,  450,  4

Okay, so we can see that there are a bunch of observations! Remember that each frame ID corresponds to 1/30 seconds. One interesting thing I can see is that the frame IDs start at 301, not 1. 

Why is that?

Since our computer vision algorithm can only run for video from the main broadcast camera, we get rid of replays or any other video that's from a different camera. My guess here is that the start of the power play was filmed with a different camera.

What do we do?

Let's look into the shot information file! In the context of the computer vision data, shots are **video shots**. A (video) shot is a continuous sequence of video filmed from **one** camera.

In [12]:
video_shot_information = pd.read_csv('2022-02-08 Canada at USA/videoShotsInfo_2022-02-08 Canada at USA P1 PP1.csv')
video_shot_information

shot_ind  frame_start_shot  frame_end_shot  time_start_shot(sec)  \
0         1               301            1070                  10.0   

   time_end_shot(sec)  
0               35.67

Let's interpret this data!

We have **one** shot from the main broadcast camera. We can see that before frame 301, the video probably came from another camera. This means that our player tracking data starts **10 seconds** into the power play. This means that the time on the clock when the tracking starts is 386 (time of the start of the power play) - 10 (time until the tracking starts) = **376 seconds**.

The tracking ends at clock seconds 386 (time of the start of the power play) - 35.67 (length of the video shot) = **350 seconds** (I just rounded to get even seconds, since that's what's in the event data).

Let's get the matching event data for this!

In [13]:
matching_event_data = pp_events = period_events.loc[(period_events['clock_seconds'] <= 376) &
                              (period_events['clock_seconds'] >= 350)]
matching_event_data.head()

game_date  season_year                 team_name  \
441  8/2/2022         2021  Olympic (Women) - Canada   
442  8/2/2022         2021  Olympic (Women) - Canada   
443  8/2/2022         2021  Olympic (Women) - Canada   
444  8/2/2022         2021  Olympic (Women) - Canada   
445  8/2/2022         2021  Olympic (Women) - Canada   

                       opp_team_name venue  period  clock_seconds  \
441  Olympic (Women) - United States  away       1            375   
442  Olympic (Women) - United States  away       1            373   
443  Olympic (Women) - United States  away       1            366   
444  Olympic (Women) - United States  away       1            365   
445  Olympic (Women) - United States  away       1            363   

    situation_type  goals_for  goals_against  ... event_successful x_coord  \
441         5 on 4          0              0  ...                t     185   
442         5 on 4          0              0  ...                t     161   
443         5 on 4          0              0  ...                t     176   
444         5 on 4          0              0  ...                t     130   
445         5 on 4          0              0  ...                t     139   

    y_coord  event_type        player_name_2 x_coord_2 y_coord_2  \
441       7      Direct  Marie-Philip Poulin     139.0       5.0   
442       3      Direct     Rebecca Johnston     143.0      19.0   
443       3      Direct  Marie-Philip Poulin     130.0      17.0   
444      17      Direct         Erin Ambrose     133.0      55.0   
445      55      Direct       Brianne Jenner     176.0      42.0   

     event_detail_1  event_detail_2 event_detail_3  
441             NaN             NaN            NaN  
442             NaN             NaN            NaN  
443             NaN             NaN            NaN  
444             NaN             NaN            NaN  
445             NaN             NaN            NaN  

[5 rows x 22 columns]

And that's it! I'm really looking forward to the interesting projects that you come up with!

Please feel free to connect with me during office hours if you have any further questions.